sycl/doc/design/MappingHostAddressesToDeviceEntities.md

```diff
@@ -0,0 +1,344 @@
+  problems with gdb `l` command and code coverage tools
+- checksum of host and device source files becomes different which causes device
+  code debugging to be completely broken in some environments (such as MS Visual
+  Studio, for example)
+
+Customizing host compiler allows to avoid issues with debuggers and code
+coverage tools, but that is not an option if a user wants to compile host part
+of an app with a 3rd-party host compiler.
+
+Further sections describe the implementation design of both approaches in more
+details, note that there are few components which should be modified regardless
+of which approach is in use.
+
+## Common front-end part
+
+DPC++ FE should support the following attribute:
+`[[__sycl_detail__::uniquely_identifiable_object(kind)]]`. This attribute accepts
+a string literal and should be applied to types (like `device_global` or
+`specialization_id`).
+
+Presence of the attribute instructs the compiler to perform the following
+things:
+- emit `sycl-unique-id` LLVM IR attribute on each definition of a variable of
+  type marked with `[[__sycl_detail__::uniquely_identifiable_object(kind)]]`
+  attribute. `sycl-unique-id` LLVM IR attribute should be accompanied by a
+  unique string identifier of a variable it is attached to. The rules for
+  creating this string are the same as for `__builtin_sycl_unique_stable_id` and
+  the same algorithm can be used when generating the string for the attribute
+- emit `sycl-uid-kind` LLVM IR attribute alongside `sycl-unique-id`, which
+  contains the `kind` string passed to
+  `[[__sycl_detail__::uniquely_identifiable_object(kind)]]` attribute
+
+**TODO**: we have `[[__sycl_detail__::device_global]]` attribute documented in
+[device global design doc][device-global-design], which instructs front-end to
+emit some additional semantic checking. Shall we leave it in place or that
+request for semantic checking should also be documented by
+`[[__sycl_detail__::uniquely_identifiable_object(kind)]]` attribute when `kind`
+is set to a certain value?
+
+**TODO**: alternatively, we could completely re-use existing
+`[[__sycl_detail__::device_global]]` attribute and introduce another one for
+specialization constants, i.e. it is a question of whether or not we want to
+generalize unique IDs generation in form of a generic attribute or not.
+
+When DPC++ compiler is used as both host and device compiler, then the attribute
+should be respected by both host and device compiler passes and LLVM IR
+attributes should appear in LLVM IR for both host and device code. When DPC++
+compiler is only used as a device compiler, then we don't expect the attribute
+to be handled on host, apparently.
+
+Another thing we need from DPC++ FE host compiler is to define a special macro,
+which will allow to distinguish it from other host compilers. That is needed to
+apply the aforementioned attribute conditionally to avoid spamming users with
```

Comment:
Can you explain this further? Won't this issue exist for all attributes in headers at the moment - since none will be recognized by custom host compiler? Anyway these attributes will be used in headers right? I need to verify this but I believe we suppress warnings from headers. @Fznamznon do you know?

Reply:
> Can you explain this further? Won't this issue exist for all attributes in headers at the moment - since none will be recognized by 3rd party host compiler?

Yes, the issue will exists for all of the existing attributes, but they can be split into two groups:
- non-user-visible ones, which we set internally in our headers: AFAIK, all of them are for device code (i.e. guarded by `#ifdef __SYCL_DEVICE_ONLY__` and the only exception is the attribute I'm proposing here
- user-visible ones, which were explicitly set by users - for them warnings from host compiler are expected, I suppose

> I need to verify this but I believe we suppress warnings from headers. 

AFAIK, we treat `sycl.hpp` as a system include, which suppresses all warnings coming out of it

Reply:
>guarded by #ifdef __SYCL_DEVICE_ONLY__ and the only exception is the attribute I'm proposing here

OK. Makes sense

>Not sure that other compilers do that though.

Oh right.  My bad.